hash-tables/problem-set-hash-tables.md

```diff
@@ -0,0 +1,287 @@
+            frequency_hash[char] += 1
+        else:
+            frequency_hash[char] = 1
+    return frequency_hash
+
+
+def is_permutation(red, blue):
+    red_frequency_hash = populate_frequency_map(red)
+    blue_frequency_hash = populate_frequency_map(blue)
+
+    return red_frequency_hash == blue_frequency_hash
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 3 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: 9LkdRV
+* title: Hash Tables
+* points: 3
+### !question
+
+Write a function `is_palindrome_permutation` that takes a string as an argument and returns `True` if the letters could be re-arranged into a palindrome.
+
+Here are the tests:
+
+```python
+def test_is_palindrome_permutation_simple():
+    assert is_palindrome_permutation("lool")
+
+
+def test_is_palindrome_permutation_palindrome():
+    assert is_palindrome_permutation("racecar")
+
+
+def test_is_palindrome_permutation_rearranged_palindrome():
+    assert is_palindrome_permutation("carrace")
+
+
+def test_is_palindrome_permutation_incomplete():
+    assert not is_palindrome_permutation("raceca")
+
+
+def test_is_palindrome_permutation_unmatching_letters():
+    assert not is_palindrome_permutation("hello")
+
+
+def test_is_palindrome_permutation_empty_string():
+    assert is_palindrome_permutation("")
+```
+
+### !end-question
+### !tests
+```python
+import unittest
+from main import *
+class TestChallenge(unittest.TestCase):
+    # Note: Written with these asserts to mimic the pytest assertions we give students
+
+    def test_is_palindrome_permutation_simple(self):
+        self.assertTrue(is_palindrome_permutation("lool"))
+
+    def test_is_palindrome_permutation_palindrome(self):
+        self.assertTrue(is_palindrome_permutation("racecar"))
+
+    def test_is_palindrome_permutation_rearranged_palindrome(self):
+        self.assertTrue(is_palindrome_permutation("carrace"))
+
+    def test_is_palindrome_permutation_incomplete(self):
+        self.assertTrue(not is_palindrome_permutation("raceca"))
+
+    def test_is_palindrome_permutation_unmatching_letters(self):
+        self.assertTrue(not is_palindrome_permutation("hello"))
+
+    def test_is_palindrome_permutation_empty_string(self):
+        self.assertTrue(is_palindrome_permutation(""))
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def is_palindrome_permutation(text):
+    frequency_hash = {}
+    for char in text:
+        if frequency_hash.get(char):
+            frequency_hash[char] += 1
+        else:
+            frequency_hash[char] = 1
+
+    num_of_single_matches_allowed = 1 if len(text) % 2 == 1 else 0
+
+    for char, frequency in frequency_hash.items():
+        if frequency % 2 != 0 and num_of_single_matches_allowed:
+            num_of_single_matches_allowed -= 1
+        elif frequency % 2 != 0:
+            return False
+
+    return True
```

Comment:
Alternatively, I like a slightly different approach that counts the number of odd matches, and compares them to the number of allowed odd matches (1). No compound conditional or elif required.
```suggestion
    odd_matches = 0
    for frequency in frequency_hash.values():
        if frequency % 2:
            odd_matches += 1
            
    return odd_matches <= 1
```